src/utils/text-measurement.js

```diff
@@ -76,14 +76,34 @@ class TextMeasurement {
 
     // Approximate the number of characters to truncate to,
     // using avg character width as reference.
-    const f = (maxWidth - this.minWidth) / this._averageCharWidth;
+    const availableWidth = maxWidth - this.minWidth;
+    const f = availableWidth / this._averageCharWidth;
     let n = Math.floor(f);
-    if (n === f) {
-      // The approximate width of `n` characters is exactly max width,
-      // so take one character less just in case.
-      n -= 1;
+    if (n < 1) {
+      return '';
     }
-    return n > 0 ? text.substring(0, n) + this.overflowChar : '';
+
+    const getWidth = () => this.getTextWidth(text.substring(0, n));
+    const getRemainingCharacterCount = () =>
+      Math.round((availableWidth - getWidth()) / this._averageCharWidth);
```

Comment:
Could you move these functions above the definition of the `n` variable, and make them take `n` as an argument? That will make it easier to understand how the loop works. I find the current style confusing, because we're touching a local variable and it changes the return value of a function which doesn't obviously take the changed variable as an input.